File: src/api/mcp-endpoint.ts
import { invoke } from "./gesture";

/**
 * The MCP endpoint's listener: which port it binds, and whether it is listening (see
 * docs/spec/mcp-server.md, "Address").
 *
 * The port is the user's setting unless `ARLESH_MCP_PORT` overrides it. A port that cannot be bound
 * — most often because another Arlesh holds it — is reported as `failed` with the reason, and a
 * restart or a new port tries again without restarting the app.
 */

/** What the listener is doing. */
export type McpListenerState =
  | { state: "starting" }
  | { state: "listening"; address: string }
  | { state: "failed"; reason: string };

/** The endpoint as the settings page shows it. */
export interface McpEndpointStatus {
  listener: McpListenerState;
  /** The port the listener last tried: the override when set, else the setting. */
  port: number;
  /** The port the user chose, or the default. */
  configured_port: number;
  /** The port `ARLESH_MCP_PORT` sets, which wins over the setting while it is set. */
  env_override: number | null;
}

/** The listener as it stands. */
export async function fetchMcpEndpointStatus(): Promise<McpEndpointStatus> {
  return invoke<McpEndpointStatus>("mcp_endpoint_status");
}

/** Binds the listener again on the current port. A failure is in the status, not thrown. */
export async function restartMcpEndpoint(): Promise<McpEndpointStatus> {
  return invoke<McpEndpointStatus>("restart_mcp_endpoint");
}

/** Saves `port` as the MCP port and moves the listener onto it. */
export async function setMcpPort(port: number): Promise<McpEndpointStatus> {
  return invoke<McpEndpointStatus>("set_mcp_port", { port });
}

File: src-tauri/src/commands/mcp_endpoint.rs
//! Tauri commands for the MCP endpoint's listener: its status, a restart, and its port.

use tauri::State;

use crate::{
    error::WireError,
    mcp::endpoint::{EndpointError, EndpointStatus, McpEndpoint},
};

/// The listener as it stands: listening on which address, or failed and why.
#[tauri::command]
pub async fn mcp_endpoint_status(
    endpoint: State<'_, McpEndpoint>,
) -> Result<EndpointStatus, WireError> {
    Ok(endpoint.status().await)
}

/// Binds the listener again on the current port. A bind failure is in the status, not an error.
#[tauri::command]
pub async fn restart_mcp_endpoint(
    endpoint: State<'_, McpEndpoint>,
) -> Result<EndpointStatus, WireError> {
    Ok(endpoint.restart().await)
}

/// Saves `port` as the MCP port and moves the listener onto it.
#[tauri::command]
pub async fn set_mcp_port(
    endpoint: State<'_, McpEndpoint>,
    port: u16,
) -> Result<EndpointStatus, WireError> {
    endpoint.set_port(port).await.map_err(|error| match error {
        EndpointError::InvalidPort => WireError::invalid_request(error.to_string()),
        EndpointError::Save(_) => WireError::internal(error.to_string()),
    })
}
